### DISCUSSION TITLE
Looking for a way to make DHCP server in softAP mode for esp32

### DISCUSSION THREAD

**Discussion by lsd-techno** [nanoframework/Home#932] [Q&A] [answered]
I've managed to migrate WiFiAP sample from windows.devices.wifi to system.device.wifi to use on esp32-pico firmware.
Everything works well with softAP and HTTPlistener but I can not figure out how I can assign IP address to connected wifi station.

Appreciate any suggestion.

**Comment by josesimoes** [comment]
To set a fixed IP configuration you have to use this [method](https://docs.nanoframework.net/api/nanoFramework.Networking.WiFiNetworkHelper.html#nanoFramework_Networking_WiFiNetworkHelper_ConnectFixAddress_System_String_System_String_nanoFramework_Networking_IPConfiguration_System_Device_WiFi_WiFiReconnectionKind_System_Boolean_System_Int32_System_Threading_CancellationToken_)

**Reply by lsd-techno** [reply]
Looks like WiFiNetworkHelper.ConnectFixAddress method is for connecting to another AP.
And I'm running esp32 in AP mode:

```
      string SoftApIP = "192.168.4.1";
      
      NetworkInterface ni = GetInterface();
      WirelessAPConfiguration wapconf = GetConfiguration();
      
      ni.EnableStaticIPv4(SoftApIP, "255.255.255.0", SoftApIP);
      wapconf.Options = WirelessAPConfiguration.ConfigurationOptions.AutoStart |
                                  WirelessAPConfiguration.ConfigurationOptions.Enable;
      wapconf.MaxConnections = 3;
      wapconf.Authentication = AuthenticationType.Open;
      wapconf.Password = "";
      wapconf.SaveConfiguration();
      
      //reboot();

```

Then I can intercept client/station connect/disconnect event... 
`
      NetworkChange.NetworkAPStationChanged += NetworkChange_NetworkAPStationChanged;
`

```
        private static void NetworkChange_NetworkAPStationChanged(int NetworkIndex, NetworkAPStationEventArgs e)
        {
            Debug.WriteLine($"NetworkAPStationChanged event Index:{NetworkIndex} Connected:{e.IsConnected} Station:{e.StationIndex} ");

            // if connected then get information on the connecting station 
            if (e.IsConnected)
            {
                WirelessAPConfiguration wapconf = WirelessAPConfiguration.GetAllWirelessAPConfigurations()[0];
                WirelessAPStation station = wapconf.GetConnectedStations(e.StationIndex);

                string macString = BitConverter.ToString(station.MacAddress);
                Debug.WriteLine($"Station mac {macString} Rssi:{station.Rssi} PhyMode:{station.PhyModes} ");

                connectedCount++;

                // Start web server when it connects otherwise the bind to network will fail as 
                // no connected network. Start web server when first station connects 
                if (connectedCount == 1)
                {
                    // Wait for Station to be fully connected before starting web server
                    // other you will get a Network error
                    Thread.Sleep(2000);
                    server.Start();
                }
            }
            else
            {
                // Station disconnected. When no more station connected then stop web server
                if (connectedCount > 0)
                {
                    connectedCount--;
                    if (connectedCount == 0)
                        server.Stop();
                }
            }
        }

```

This part works well but when I connect to my softAP with mobile phone it never gets IP address from AP.
Only if I assign IP manually on the phone then I can access to WebServer running on the esp32.

On connection I can see MAC address of connected device... that would be best if I can assign IP address based on connected devices' MAC.

**Comment by lsd-techno** [comment, marked as answer]
I did some dirty implementation of DHCP server [here](https://github.com/lsd-techno/WiFiAP)

That solve my problem to avoid manual setting of IP address on every device that I want to connect to nanoframework device in AP mode.

That would be good to make some nuget package with better implementation, but i'm not good enough to make that.

**Comment by josesimoes** [comment]
@lsd-techno that's great that you've managed to find and implement a workaround. Agreed that the minimal DHCP server should be made available as a helper NuGet. 
Can you please extract this to a separate project so we can evaluate what's the best place to have it?

**Comment by sazanof** [comment]
I think, that in older versions of Core (10.x.x) DHCP works correctly (i have one board with older Core version).
Maybe this is a bug?
When I connected at ESP32 and see this IP in ipconfig command: 169.254.129.67
That means, that DHCP server not working.

**Reply by Ellerbach** [reply]
> When I connected at ESP32 and see this IP in ipconfig command: 169.254.129.67

this is a private non DHCP address. When no DHCP is present, those addresses are allocated by devices themselves.
